File: PracticasPDOORuby/Practica4_Ruby/lib/Irrgarten/combat_element.rb
# frozen_string_literal: true

module Irrgarten
  # Clase padre para Weapon y Shield
  class CombatElement
    # 'effect' generaliza a power/protection
    # 'uses' es común a ambos
    attr_reader :effect, :uses

    def initialize(effect, uses)
      @effect = effect
      @uses = uses
    end

    # Metodo que produce el efecto (atacar o proteger)
    # Devuelve el valor del efecto si hay usos, o 0 si no los hay
    def produce_effect
      if @uses > 0
        @uses -= 1
        @effect
      else
        0.0
      end
    end

    # Metodo para descartar el elemento (usar Dice)
    def discard
      Dice.discard_element(@uses)
    end

    def to_s
      "[#{@effect}, #{@uses}]"
    end
  end
end

File: PracticasPDOOJava/Practica5_Java/src/irrgarten/CombatElement.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author Joaquín Cruz Lorenzo
 */
public abstract class CombatElement {
    // Atributos
    private float effect;
    private int uses;

    // Constructor.
    public CombatElement(float effect, int uses) {
        this.effect = effect;
        this.uses = uses;
    }
    
    protected float produceEffect() {
        if (getUses() > 0) {
            this.uses--;
            return getEffect();
        } else return 0;   
    }

    /**
     * Devuelve el efecto base del elemento.
     * (Protected para que solo las clases hijas puedan verlo)
     * @return el valor del efecto
     */
    public float getEffect() {
        return effect;
    }

    /**
     * Devuelve los usos restantes.
     * (Protected para que solo las clases hijas puedan verlo)
     * @return los usos restantes
     */
    public int getUses() {
        return uses;
    }
    
    public boolean discard() {
        return Dice.discardElement(this.uses);
    }
    
    @Override
    public abstract String toString();
}
